File: app/src/main/java/ru/santaev/clipboardtranslator/util/RxHelper.java
package ru.santaev.clipboardtranslator.util;

import java.util.concurrent.Callable;

import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Action;
import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;


public class RxHelper {

    private static <T> Observable<T> getObservable(Callable<T> function) {
        return Observable.create(
                (subscriber -> {
                    T resp = function.call();
                    if (resp == null) {
                        subscriber.onError(new Exception("Error while call function " + function));
                    } else {
                        subscriber.onNext(resp);
                    }
                })
        );
    }

    public static <T> Disposable runOnIoThread(Callable<T> function,
                                               Consumer<Throwable> handler, Consumer<T> onNext,
                                               final Action onCompleted) {
        return getObservable(function)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(t -> {
                            if (onNext != null) {
                                onNext.accept(t);
                            }
                        }, handler,
                        () -> {
                            if (onCompleted != null) {
                                onCompleted.run();
                            }
                        });
    }

}

File: app/src/main/java/ru/santaev/clipboardtranslator/util/RxHelper.kt
package ru.santaev.clipboardtranslator.util

import io.reactivex.CompletableTransformer
import io.reactivex.FlowableTransformer
import io.reactivex.ObservableTransformer
import io.reactivex.SingleTransformer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers


object RxHelper {

    fun <T> getTransformer(): ObservableTransformer<T, T> {
        return ObservableTransformer { upstream ->
            upstream.subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
        }
    }

    fun <T> getFlowableTransformer(): FlowableTransformer<T, T> {
        return FlowableTransformer { upstream ->
            upstream.subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
        }
    }

    fun <T> getSingleTransformer(): SingleTransformer<T, T> {
        return SingleTransformer { upstream ->
            upstream.subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
        }
    }

    fun getCompletableTransformer(): CompletableTransformer {
        return CompletableTransformer { upstream ->
            upstream.subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
        }
    }

}
